src/beast/beast/http/impl/write.ipp

```diff
@@ -0,0 +1,420 @@
+
+    friend
+    auto asio_handler_allocate(
+        std::size_t size, write_op* op)
+    {
+        return boost_asio_handler_alloc_helpers::
+            allocate(size, op->d_->h);
+    }
+
+    friend
+    auto asio_handler_deallocate(
+        void* p, std::size_t size, write_op* op)
+    {
+        return boost_asio_handler_alloc_helpers::
+            deallocate(p, size, op->d_->h);
+    }
+
+    friend
+    auto asio_handler_is_continuation(write_op* op)
+    {
+        return op->d_->state >= 1 ||
+            boost_asio_handler_cont_helpers::
+                is_continuation(op->d_->h);
+    }
+
+    template <class Function>
+    friend
+    auto asio_handler_invoke(Function&& f, write_op* op)
+    {
+        return boost_asio_handler_invoke_helpers::
+            invoke(f, op->d_->h);
+    }
+};
+
+template<class Stream, class Message, class Handler>
+void
+write_op<Stream, Message, Handler, false>::
+    operator()(boost::system::error_code ec, std::size_t bytes_transferred)
+{
+    auto& d = *d_;
+    while(! ec && d.state != 99)
+    {
+        switch(d.state)
+        {
+        case 0:
+        case 1:
+        {
+            auto const result =
+                d.w.prepare(std::move(d.copy), ec);
+            if(ec)
+            {
+                // call handler
+                d.state = 99;
+                break;
+            }
+            if(boost::indeterminate(result))
+            {
+                // suspend
+                d.state = 0;
+                d.copy = d.resume;
+                return;
+            }
+            if(! result)
+            {
+                // call handler
+                d.state = 99;
+                break;
+            }
+            d.state = 1;
+            boost::asio::async_write(d.s,
+                d.w.data(), std::move(*this));
+            return ;
+        }
+        }
+    }
+    d.h(ec);
+    d.resume = {};
+    d.clear = {};
+}
+
+} // detail
+
+//------------------------------------------------------------------------------
+
+namespace detail {
+
+template<class SyncWriteStream,
+    bool isRequest, class Body, class Allocator>
+void
+write(SyncWriteStream& stream,
+    message<isRequest, Body, Allocator> const& m,
+        boost::system::error_code& ec, std::true_type)
+{
+    typename Body::writer w(m);
+    boost::asio::write(stream, w.data(), ec);
+}
+
+template<class SyncWriteStream,
+    bool isRequest, class Body, class Allocator>
+void
+write(SyncWriteStream& stream,
+    message<isRequest, Body, Allocator> const& msg,
+        boost::system::error_code& ec, std::false_type)
+{
+    std::mutex m;
+    std::condition_variable cv;
+    typename Body::writer w(msg);
+    bool ready = false;
+    resume_context resume{
+        [&]
+        {
+            std::lock_guard<std::mutex> lock(m);
+            ready = true;
+            cv.notify_one();
+        }};
+    auto copy = resume;
+    for(;;)
+    {
+        auto result = w.prepare(std::move(copy), ec);
```

Comment:
I don't think there's any point to the `std::move` here.  `resume` is just a type full of references, and so its move is the same as a copy.  So you can just get rid of `copy` and use `resume`.